### ISSUE TITLE
uClibc-ng 1.0.34 broken on Linux kernel 4.19.136

### ISSUE THREAD

**Issue by abrodkin** [foss-for-synopsys-dwc-arc-processors/toolchain#298]
As reported by @wbx-github with uClibc-ng 1.0.34 on Linux kernel 4.19.136, gcc 9.3.0 and binutils 2.35.
Platform used initially for reproduction is HSDK, though it worth testing on nSIM as well.
```
[    9.667903] Starting Linux (built with OpenADK).
[    9.673796] Freeing unused kernel memory: 592K
[    9.678323] This architecture does not have kernel memory
protection.
[    9.684992] Run /sbin/init as init process
[    9.693619] potentially unexpected fatal signal 11.
[    9.698600] Path: /bin/busybox
[    9.701967] CPU: 2 PID: 1 Comm: init Not tainted 4.19.136-1 #2
[    9.707882]
[    9.707882] [ECR   ]: 0x00050200 => Invalid Write @ 0x00000000 by
insn @ 0x2000181a
[    9.717373] [EFA   ]: 0x00000000
[    9.717373] [BLINK ]: 0x20001792
[    9.717373] [ERET  ]: 0x2000181a
[    9.727392]     @off 0x181a in [/lib/ld-uClibc-1.0.34.so]
[    9.727392]     VMA: 0x20000000 to 0x20008000
[    9.737355] [STAT32]: 0x80081082 : IE U    
[    9.741969] BTA: 0x20001792   SP: 0x5f8bdb9c  FP: 0x5f8bdd2c
[    9.747648] LPS: 0x00000010  LPE: 0x00000080 LPC: 0x00000000
[    9.753784] r00: 0x20005b64  r01: 0x00000025 r02: 0xffffffff
[    9.753784] r03: 0x00000022  r04: 0x00000000 r05: 0x00000000
[    9.753784] r06: 0x00000000  r07: 0x00000000 r08: 0x000000de
[    9.753784] r09: 0x0000000c  r10: 0x00000001 r11: 0x00000000
[    9.753784] r12: 0x20001792  r13: 0x00000000 r14: 0x20005b64
[    9.753784] r15: 0x2000a070  r16: 0x5f8bddf8 r17: 0x5f8bdf2c
[    9.753784] r18: 0x2000a070  r19: 0x00000000 r20: 0x5f8bddf8
[    9.753784] r21: 0x00000000  r22: 0x00000000 r23: 0x00000000
[    9.753784] r24: 0x5f8bdc18  r25: 0x00000000
[    9.753784]
[    9.753784]
[    9.807098] Kernel panic - not syncing: Attempted to kill init!
exitcode=0x0000000b
[    9.807098]
```

With kernel 5.4.53 it just hangs on this line here:
```
[    2.579235] NET: Registered protocol family 1
```

Attached is initramfs used (originally from https://debug.openadk.org/initramfs_data.cpio.xz).
[initramfs_data.cpio.gz](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/5158394/initramfs_data.cpio.gz)

**Comment by wbx-github** [comment]
Seems no toolchain issue so far, but a kernel issue.

This commit in linux/stable breaks it:
> commit 01b636cbc4c39c7073c02c2f1dfbe8c36c11938f                                                                                                                   
> Author: Vineet Gupta <vgupta@synopsys.com>                                                                                                                        
> Date:   Tue May 19 22:28:32 2020 -0700                                                                                                                            
>                                                                                                                                                                   
>     ARC: entry: fix potential EFA clobber when TIF_SYSCALL_TRACE                                                                                                  
>                                                                                                                                                                   
>     commit 00fdec98d9881bf5173af09aebd353ab3b9ac729 upstream.                                                                                                     
>                                                                                                                                                                   
>     Trap handler for syscall tracing reads EFA (Exception Fault Address),                                                                                         
>     in case strace wants PC of trap instruction (EFA is not part of pt_regs                                                                                       
>     as of current code).                                                                                                                                          
>                                                                                                                                                                   
>     However this EFA read is racy as it happens after dropping to pure                                                                                            
>     kernel mode (re-enabling interrupts). A taken interrupt could                                                                                                 
>     context-switch, trigger a different task's trap, clobbering EFA for this                                                                                      
>     execution context.                                                                                                                                            
>                                                                                                                                                                   
>     Fix this by reading EFA early, before re-enabling interrupts. A slight                                                                                        
>     side benefit is de-duplication of FAKE_RET_FROM_EXCPN in trap handler.                                                                                        
>     The trap handler is common to both ARCompact and ARCv2 builds too.                                                                                            
>                                                                                                                                                                   
>     This just came out of code rework/review and no real problem was reported                                                                                     
>     but is clearly a potential problem specially for strace.                                                                                                      
>                                                                                                                                                                   
>     Cc: <stable@vger.kernel.org>                                                                                                                                  
>     Signed-off-by: Vineet Gupta <vgupta@synopsys.com>                                                                                                             
>     Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

**Comment by vineetgarc** [comment]
Indeed that kernel commit is the culprit, it breaks EV_Trap in sublte way: working on fix !

**Comment by vineetgarc** [comment]
So here's the issue:

The bisected commit  [00fdec98d9881bf517](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/arch/arc/kernel/entry.S?id=00fdec98d9881bf5173af09aebd353ab3b9ac729) was merged upstream in v5.8 and also marked for stable backport.

In on itself this commit is not faulty, and that is the reason our in-house regressions didn't trip since it was merged in July. However it implicitly relies on an interim commit [68e5c6f073bcf70da5f8](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/arch/arc/include/asm/entry-arcv2.h?id=68e5c6f073bcf70da5f8eef88eb2d98f7c560bb6) **"ARC: entry: EV_Trap expects r10 (vs. r9) to have exception cause"** which was merged upstream only in v5.3 and **NOT** backported. Thus all stable kernels prior to v5.3 are broken

Conceptually the fix is easy, just backport that but it doesn't apply cleanly, due to many other interim changes and also because it needs to go in before the offending commit. Anyhow that can be solved. Can you confirm if a non-conflicting version attached here cures it for you (in my testing it does).

[0001-ARC-entry-EV_Trap-expects-r10-vs.-r9-to-have-excepti.patch.txt](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/5404343/0001-ARC-entry-EV_Trap-expects-r10-vs.-r9-to-have-excepti.patch.txt)

**Comment by vineetgarc** [comment]
To capture the technical details: the issue was Low level trap handler, which is the entry point for syscall and breakpoint Traps, getting borked due to register r9 getting clobbered. r9 has ECR (Exception Cause Register) used to decide to whether to dispatch trap as syscall or breakpoint. Due to the bug all syscalls were erroneously handled as breakpoints.

**Orig Code**
```
ENTRY(EV_Trap)
	EXCEPTION_PROLOGUE         # sets ECR in r9
  	bmsk.f 0, r9, 7            # checks r9 to decide syscall vs. breakpoint
        bnz    trap_with_param
	FAKE_RET_FROM_EXCPN        # clobbered r9 OK
```
**Backport NOK**
```
ENTRY(EV_Trap)
	EXCEPTION_PROLOGUE         # sets ECR in r9
	FAKE_RET_FROM_EXCPN        # offending commit moves the macro here, clobbers r9
  	bmsk.f 0, r9, 7            # check for syscall vs breakpoint is broken NOK
```
**Backport + Interim change**
```
ENTRY(EV_Trap)
	EXCEPTION_PROLOGUE         # Interim commit starts using r10 for ECR
	FAKE_RET_FROM_EXCPN        # r9 clobbered doesn't matter
  	bmsk.f 0, r10, 7           # r10 checked so OK
```

**Comment by vineetgarc** [comment]
[Oct 19] email sent to stable list to apply the revert [Revert "ARC: entry: fix potential EFA clobber when TIF_SYSCALL_TRACE"](http://lists.infradead.org/pipermail/linux-snps-arc/2020-October/004348.html)
[Nov 6] Pinged stable list

**Comment by vineetgarc** [comment]
The fix is now merged mainline to stable trees: v4.19.156

2020-10-19 259041122077 Revert "ARC: entry: fix potential EFA clobber when TIF_SYSCALL_TRACE"
